packages/application-shell/src/test-utils/test-utils.spec.js

```diff
@@ -264,3 +274,117 @@ describe('router', () => {
     expect(history.location.pathname).toBe('/foo');
   });
 });
+
+describe('custom render function', () => {
+  describe('wrapper', () => {
+    describe('when provided', () => {
+      const Context = React.createContext();
+      const ProvidedWrapper = ({ children }) => (
+        <Context.Provider value="provided wrapper">{children}</Context.Provider>
+      );
+      ProvidedWrapper.propTypes = {
+        children: PropTypes.node.isRequired,
+      };
+
+      it('should renderApp merges the provided wrapper with its own wrapper', () => {
+        const TestComponent = () => {
+          // provided wrapper
+          const value = useContext(Context);
+          // own wrapper
+          useIntl();
+
+          return value;
+        };
+
+        const rendered = renderApp(<TestComponent number={1} />, {
+          wrapper: ProvidedWrapper,
+        });
+        rendered.getByText(/provided wrapper/i);
+      });
+
+      it('should renderAppWithRedux merges the provided wrapper with its own wrapper', () => {
+        const TestComponent = () => {
+          // provided wrapper
+          const value = useContext(Context);
+          // own wrapper
+          useSelector(() => {});
+
+          return value;
+        };
+
+        const rendered = renderAppWithRedux(<TestComponent number={1} />, {
+          wrapper: ProvidedWrapper,
+        });
+        rendered.getByText(/provided wrapper/i);
+      });
+    });
+
+    describe('when not provided', () => {
+      it.each([renderApp, renderAppWithRedux])(
+        'should %p still manage',
```

Comment:
I fail to grasp the description here. Maybe it's due to the structure or description. Wondering if a structure as in

```txt
describe: when rendering with redux
   describe: with wrapper
       it: should render and merge the passed and internal wrapper
       it: should allow rerendering the passed component
   describe: without wrapper
       it: should render and internal wrapper
       it: should allow rerendering the passed component
   describe: with experimental rendering
       it: should render and internal wrapper
       it: should allow rerendering the passed component
describe with rendering without redux
   describe: with wrapper
       it: should render and merge the passed and internal wrapper
       it: should allow rerendering the passed component
   describe: without wrapper
       it: should render and internal wrapper
       it: should allow rerendering the passed component
```

Reply:
After implementing [this suggestion](https://github.com/commercetools/merchant-center-application-kit/pull/1165#discussion_r346733517), the structure becomes

```
  custom render functions
    with wrapper
      ✓ should merge the passed wrapper with renderApp internal wrapper (6ms)
      ✓ should merge the passed wrapper with renderAppWithRedux internal wrapper (11ms)
    without wrapper
      ✓ should [Function renderApp] still manage (7ms)
      ✓ should [Function renderAppWithRedux] still manage (6ms)
    rerender
      ✓ should work with renderApp (5ms)
      ✓ should work with renderAppWithRedux (10ms)
      ✓ should work with experimentalRenderAppWithRedux (9ms)
```

*still manage* is not clear enough and I will replace it by *still work*. This test is meant to ensure that app-kit custom render functions will still work in case a `wrapper` wasn't provided.

Reply:
Looks way better. 

1. [Function renderApp] looks odd obvously 

Super nitty now and sorry about it

```txt
describe: rendering
  it: should be supported by renderApp
  ...
```

Reply:
No worries. The thing with `[Function renderApp]` is that it was the best I can do when using `it.each`.

https://github.com/commercetools/merchant-center-application-kit/blob/02eccdbe6808552508d0213bb18499a72d6cb42d/packages/application-shell/src/test-utils/test-utils.spec.js#L322-L330

I needed to pass the actual functions in the array and not as strings to be able to use them here.

https://github.com/commercetools/merchant-center-application-kit/blob/02eccdbe6808552508d0213bb18499a72d6cb42d/packages/application-shell/src/test-utils/test-utils.spec.js#L327

`%p` is the best formatting that I found. In the end, I preferred having `[Function renderApp]`with a single test and avoid duplication rather than write two explicit tests. It's true still that it looks odd but I find it readable at the same time.

> should be supported by renderApp

doesn't reflect clearly that we're guarding against issues when rendering without providing a wrapper. The reason why I added this test is because of the `wrapIfNeeded` function.

https://github.com/commercetools/merchant-center-application-kit/blob/02eccdbe6808552508d0213bb18499a72d6cb42d/packages/application-shell/src/test-utils/test-utils.js#L237-L238

In the beginning, I implemented it without guarding against the `wrapper` being undefined. This broke the rendering logic and none of the tests available back then caught it.

Reply:
Ah got it. Thanks for the explanation!